include/mesh.tcc

```diff
@@ -1885,9 +1920,419 @@ void mpm::Mesh<Tdim>::create_nodal_properties() {
   }
 }
 
+//! Compute free surface cells, nodes, and particles
+template <unsigned Tdim>
+bool mpm::Mesh<Tdim>::compute_free_surface(std::string method,
+                                           double volume_tolerance) {
+  if (method == "density") {
+    return this->compute_free_surface_by_density(volume_tolerance);
+  } else if (method == "geometry") {
+    return this->compute_free_surface_by_geometry(volume_tolerance);
+  } else {
+    console_->info(
+        "The selected free-surface computation method: {}\n is not available. "
+        "Using density approach as default method.",
+        method);
+    return this->compute_free_surface_by_density(volume_tolerance);
+  }
+}
+
+//! Compute free surface cells, nodes, and particles by density and geometry
+template <unsigned Tdim>
+bool mpm::Mesh<Tdim>::compute_free_surface_by_geometry(
+    double volume_tolerance) {
+  bool status = true;
+  try {
```

Comment:
If we move some of the functions below outside of this function, we can have this function outside mesh.